### PR TITLE
Add the support of bypassing JWT authn for CORS requests

### LINKED ISSUE #651: JWT-AUTH and OPTIONS requests
Hi,

I started to play with the `jwt-auth` and it works well, I could even integrate it with my Keycloak IDP. But now I'm trying to do a request from the browser and it fails on the `OPTIONS` preflight because the authorization header is not present (the browser remove it automatically). Is there an easy way to tell the filter to skip 'OPTIONS' requests (even if it's a bit hacky, it's for demo purposes) ?

Thx

### PR DESCRIPTION
**Pull request by lei-tang**
**What this PR does / why we need it**: Add the support of bypassing JWT authn for CORS requests

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #651 

**Special notes for your reviewer**:

**Release note**:

```release-note
```

### CODE REVIEW

File: src/envoy/http/jwt_auth/jwt_authenticator.cc [hunk 1]
```diff
@@ -214,11 +217,20 @@ bool JwtAuthenticator::OkToBypass() {
     return true;
   }
 
+  // Per the spec
+  // http://www.w3.org/TR/cors/#cross-origin-request-with-preflight-0, CORS
+  // pre-flight requests shouldn't include user credentials.
+  if (headers_->Method() &&
```
Comment (qiwzhang) [inline src/envoy/http/jwt_auth/jwt_authenticator.cc]: how about bail out earlier, like in line 63?

Comment (lei-tang) [reply src/envoy/http/jwt_auth/jwt_authenticator.cc]: Done.

File: src/envoy/http/jwt_auth/jwt_authenticator.cc [hunk 2]
```diff
@@ -62,6 +62,17 @@ void JwtAuthenticator::Verify(HeaderMap &headers,
   headers_ = &headers;
   callback_ = callback;
 
+  // Per the spec
+  // http://www.w3.org/TR/cors/#cross-origin-request-with-preflight-0, CORS
+  // pre-flight requests shouldn't include user credentials.
+  if (headers_->Method() &&
+      LowerCaseString(kOptionsHttpMethod) ==
```
Comment (qiwzhang) [inline src/envoy/http/jwt_auth/jwt_authenticator.cc]: I think Envoy has fixed string for OPTIONS

https://github.com/envoyproxy/envoy/blob/master/source/common/http/headers.h#L176

Comment (qiwzhang) [reply src/envoy/http/jwt_auth/jwt_authenticator.cc]: I prefer to do this check in Fitler.cc right inside the decodeHeaders().

**Comment by qiwzhang** [comment]
Could you help to add this for Envoy one too?

It is here

https://github.com/envoyproxy/envoy/blob/master/source/extensions/filters/http/jwt_authn/filter.cc#L23

File: src/envoy/http/jwt_auth/jwt_authenticator.cc (hunk 2 above)
Comment (lei-tang) [reply src/envoy/http/jwt_auth/jwt_authenticator.cc]: decodeHeaders() does not have tests. Placing the check in Verify() will enable tests. Plus, placing the check in Verify() is equivalent to placing the check in decodeHeaders().

**Comment by lei-tang** [comment]
Sure, I can add this PR to Envoy repo after this PR is merged.

File: src/envoy/http/jwt_auth/jwt_authenticator.cc (hunk 2 above)
Comment (lei-tang) [reply src/envoy/http/jwt_auth/jwt_authenticator.cc]: Done: changed to use the OPTIONS fixed string from Envoy.

File: src/envoy/http/jwt_auth/jwt_authenticator.cc [hunk 3]
```diff
@@ -66,9 +63,9 @@ void JwtAuthenticator::Verify(HeaderMap &headers,
   // http://www.w3.org/TR/cors/#cross-origin-request-with-preflight-0, CORS
   // pre-flight requests shouldn't include user credentials.
   if (headers_->Method() &&
-      LowerCaseString(kOptionsHttpMethod) ==
+      LowerCaseString(Http::Headers::get().MethodValues.Options) ==
```
Comment (qiwzhang) [inline src/envoy/http/jwt_auth/jwt_authenticator.cc]: Why do we need to lower case the method?   I thought http methods are pre-defined, they are always upper case.

Comment (lei-tang) [reply src/envoy/http/jwt_auth/jwt_authenticator.cc]: Done: lowercase method has been removed.

File: src/envoy/http/jwt_auth/jwt_authenticator.cc [hunk 4]
```diff
@@ -49,16 +49,26 @@ void ExtractUriHostPath(const std::string& uri, std::string* host,
...
-                              JwtAuthenticator::Callbacks* callback) {
+void JwtAuthenticator::Verify(HeaderMap &headers,
+                              JwtAuthenticator::Callbacks *callback) {
   headers_ = &headers;
   callback_ = callback;
 
+  // Per the spec
+  // http://www.w3.org/TR/cors/#cross-origin-request-with-preflight-0, CORS
+  // pre-flight requests shouldn't include user credentials.
+  if (headers_->Method() && Http::Headers::get().MethodValues.Options ==
```
Comment (lizan) [inline src/envoy/http/jwt_auth/jwt_authenticator.cc]: This bypasses too many requests, e.g. WebDAV uses OPTIONS for different purpose and those shouldn't be bypassed.

Comment (lei-tang) [reply src/envoy/http/jwt_auth/jwt_authenticator.cc]: https://github.com/istio/proxy/pull/2140 is created to add more checks for CORS requests.

**Pull request merged**